controllers/configurationpolicy_utils.go

```diff
@@ -577,98 +577,161 @@ func createStatus(
-			case reasonDeleteSuccess:
-				generatedReason = reasonDeleteSuccess
-				generatedMsg = fmt.Sprintf("%s%s was deleted successfully", resourceName, namesStr)
-			case reasonWantFoundDNE:
-				generatedReason = "K8s does not have a `must have` object"
-				compliancyDetailsMsg += fmt.Sprintf("%s%s not found", resourceName, namesStr)
-			case reasonWantFoundNoMatch:
-				generatedReason = "K8s does not have a `must have` object"
-				compliancyDetailsMsg += fmt.Sprintf("%s%s found but not as specified", resourceName, namesStr)
-			case reasonWantNotFoundExists:
-				generatedReason = "K8s has a `must not have` object"
-				compliancyDetailsMsg += fmt.Sprintf("%s%s found", resourceName, namesStr)
-			case reasonWantNotFoundDNE:
-				generatedReason = "K8s `must not have` object already missing"
-				compliancyDetailsMsg += fmt.Sprintf("%s%s missing as expected", resourceName, namesStr)
-			default:
-				// If it's not one of the above reasons, then skip consolidation. This is likely an error being
-				// reported.
-				if i == 0 {
-					if compliancyDetailsReason != "" {
-						compliancyDetailsReason += "; "
-					}
-
-					compliancyDetailsReason += reason
-				}
+		msgMap, compliancyDetailsReason, compliancyDetailsMsg = generateMessageWithReason(sortedObjectNamesStrs,
+			objectNameStrsToNsNames, nsNameToEvent,
+			compliancyDetailsReason, compliancyDetailsMsg, reason)
 
-				for j, nsName := range objectNameStrsToNsNames[namesStr] {
-					if j != 0 {
-						compliancyDetailsMsg += "; "
-					}
+		compliancyDetailsMsg = getCombinedCompliancyDetailsMsg(msgMap, resourceName, compliancyDetailsMsg)
+	}
 
-					compliancyDetailsMsg += nsNameToEvent[nsName].event.message
-				}
+	return
+}
 
-				// Assume the included messages include the namespace.
-				continue
-			}
+func setCompliancyDetailsMsgEnd(compliancyDetailsMsg string) string {
+	if compliancyDetailsMsg != "" && !strings.HasSuffix(compliancyDetailsMsg, "; ") {
+		compliancyDetailsMsg += "; "
+	}
+
+	return compliancyDetailsMsg
+}
+
+func getCombinedCompliancyDetailsMsg(msgMap map[string][]string,
+	resourceName string, compliancyDetailsMsg string,
+) string {
+	// Sort msgTemplate keys for consistent processing
+	msgTemplates := make([]string, 0, len(msgMap))
+	for k := range msgMap {
+		msgTemplates = append(msgTemplates, k)
+	}
+
+	slices.Sort(msgTemplates)
+
+	for _, msgTemplate := range msgTemplates {
+		names := msgMap[msgTemplate]
+		slices.Sort(names) // Sort names for consistent output
+
+		compliancyDetailsMsg = setCompliancyDetailsMsgEnd(compliancyDetailsMsg)
+
+		joinedNames := strings.Join(names, ", ")
+		if joinedNames != "" {
+			joinedNames = fmt.Sprintf(" [%s]", joinedNames)
+		}
+
+		compliancyDetailsMsg += fmt.Sprintf(msgTemplate, resourceName, joinedNames)
+	}
+
+	return compliancyDetailsMsg
+}
 
-			// This prevents repeating the same reason for each unique object name list.
+// Process the object name strings in order to ensure a deterministic reason and message.
```

Comment:
Nit: since this function returns a map (and lookups from maps in go can be in an inconsistent order), it is the `getCombinedCompliancyDetailsMsg` and `createStatus` functions which ensures they are in a deterministic order.

Also, I just want to thank you for bringing this into a separate function! It really helps readability of `createStatus` :1st_place_medal:

Reply:
You are right, how about move "// Process the object name strings in order to ensure a deterministic reason and message." above "getCombinedCompliancyDetailsMsg" as function comment